sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/ReadSchemaTest.scala

```diff
@@ -108,6 +108,40 @@ trait AddColumnTest extends ReadSchemaTest {
   }
 }
 
+trait AddColumnIntoTheMiddleTest extends ReadSchemaTest {
+  import testImplicits._
+
+  test("append column into middle") {
+    withTempPath { dir =>
+      val path = dir.getCanonicalPath
+
+      val df1 = Seq((1, 2, "abc"), (4, 5, "def"), (8, 9, null)).toDF("col1", "col2", "col3")
+      val df2 = Seq((10, null, 20, null), (40, "uvw", 50, "xyz"), (80, null, 90, null))
+        .toDF("col1", "col4", "col2", "col3")
+
+      val dir1 = s"$path${File.separator}part=one"
+      val dir2 = s"$path${File.separator}part=two"
+
+      df1.write.format(format).options(options).save(dir1)
+      df2.write.format(format).options(options).save(dir2)
+
+      val df = spark.read
+        .schema(df2.schema)
```

Comment:
just for curiosity, if we don't specify the schema, can spark infer the schema correctly?

Reply:
If memory serves me well, schema merging is not supported for ORC. There’s an open issue for implementing it.

The current behavior is that the schema of the first file will be picked, and the rest will be ignored.

Reply:
For future reference:

- [Support merge schema for ORC](https://issues.apache.org/jira/browse/SPARK-11412)
- [Feature parity for ORC with Parquet](https://issues.apache.org/jira/browse/SPARK-20901)